fuzz/cpio_rdr.c

```diff
@@ -0,0 +1,438 @@
+    struct header_old_cpio *h = (struct header_old_cpio *)cpio_header;
+
+    if (h->c_magic == 070707)
+        return h->c_namesize;
+    return old_swap(h->c_namesize);
+}
+
+static size_t old_datasize(void *cpio_header) {
+    struct header_old_cpio *h = (struct header_old_cpio *)cpio_header;
+
+    if (h->c_magic == 070707)
+        return h->c_filesize[0] << 16 | h->c_filesize[1];
+    return old_swap(h->c_filesize[0]) << 16 | old_swap(h->c_filesize[1]);
+}
+
+static const struct cpio_reader_st cpio_old = {
+    old_headersize, old_filenamesize, old_datasize, 2,
+    NULL, NULL, NULL, NULL
+};
+
+/* Portable ASCII */
+struct cpio_odc_header {
+    char c_magic[6];
+    char c_dev[6];
+    char c_ino[6];
+    char c_mode[6];
+    char c_uid[6];
+    char c_gid[6];
+    char c_nlink[6];
+    char c_rdev[6];
+    char c_mtime[11];
+    char c_namesize[6];
+    char c_filesize[11];
+};
+
+static uint64_t octstr2uint64(const char *str, size_t len)
+{
+    uint64_t val = 0;
+
+    OPENSSL_assert(len <= 11);
+    while (len-- > 0) {
+        unsigned char c = *str++;
+
+#ifdef CHARSET_EBCDIC
+        c = os_toebcdic[c];
+#endif
+
+        switch (c) {
+        case '0':
+            c = 0;
+            break;
+        case '1':
+            c = 1;
+            break;
+        case '2':
+            c = 2;
+            break;
+        case '3':
+            c = 3;
+            break;
+        case '4':
+            c = 4;
+            break;
+        case '5':
+            c = 5;
+            break;
+        case '6':
+            c = 6;
+            break;
+        case '7':
+            c = 7;
+            break;
+        }
+        val = (val << 3) | c;
+    }
+
+    return val;
+}
+
+static size_t odc_headersize(void *cpio_header) {
+    struct cpio_odc_header *h = (struct cpio_odc_header *)cpio_header;
+
+    return
+        strncmp(h->c_magic, "070707", sizeof(h->c_magic)) == 0 ? sizeof(*h) : 0;
+}
+
+static size_t odc_filenamesize(void *cpio_header) {
+    struct cpio_odc_header *h = (struct cpio_odc_header *)cpio_header;
+
+    return octstr2uint64(h->c_namesize, sizeof(h->c_namesize));
+}
+
+static size_t odc_datasize(void *cpio_header) {
+    struct cpio_odc_header *h = (struct cpio_odc_header *)cpio_header;
+
+
+    return octstr2uint64(h->c_filesize, sizeof(h->c_filesize));
+}
+
+static const struct cpio_reader_st cpio_odc = {
+    odc_headersize, odc_filenamesize, odc_datasize, 1,
+    NULL, NULL, NULL, NULL
+};
+
+/* New ASCII format */
+struct cpio_newc_header {
+    char c_magic[6];
+    char c_ino[8];
+    char c_mode[8];
+    char c_uid[8];
+    char c_gid[8];
+    char c_nlink[8];
+    char c_mtime[8];
+    char c_filesize[8];
+    char c_devmajor[8];
+    char c_devminor[8];
+    char c_rdevmajor[8];
+    char c_rdevminor[8];
+    char c_namesize[8];
+    char c_check[8];
+};
+
+static uint32_t hexstr2uint32(const char *str, size_t len)
+{
+    uint32_t val = 0;
+
+    OPENSSL_assert(len <= 8);
+    while (len-- > 0) {
```

Comment:
`len` underflows in last iteration.

Reply:
Does it matter?  it's not used after the loop, and it's the pre-decrement value that's checked.

Reply:
No, only it may be flagged by clang's ubsan...